### ISSUE TITLE
steam-runtime's eglibc incorrectly adds leaf attribute to dlopen

### ISSUE THREAD

**Issue by rbernon** [ValveSoftware/steam-runtime#250]
As in the title, the eglibc version shipped with steam-runtime marks dlopen with leaf attribute (__THROW instead of __THROWNL). This makes GCC assume a dlopen call can have no side-effect on the current translation unit.

Wine 5.6 directly calls dlopen instead of the legacy wine_dlopen wrapper that was used before (to workaround another libc issue), and expects the call to have some side effects on the translation unit as the result of a wine builtin library load. GCC optimizes that away, because of the leaf attribute, and wine fails to load its library when built with steam-runtime.

It's probably easier to workaround that in Wine but I'm still filling this here because eglibc is obviously incorrect, and it has been fixed in later releases (https://launchpad.net/ubuntu/utopic/amd64/libc6-dev/2.19-0ubuntu6 for instance uses __THROWNL). I'm not sure how hard it would be to fix that in steam-runtime, either by updating eglibc, or just patching the header.

**Comment by smcv** [comment]
> I'm not sure how hard it would be to fix that in steam-runtime, either by updating eglibc, or just patching the header.

It should be straightforward, just very slow (eglibc takes an age to compile, even on fast autobuilders).

**Comment by smcv** [comment]
I'm guessing what's wanted here is a backport of this commit?

https://repo.or.cz/glibc.git/commitdiff/3b813b29653a3ebfd78c2529127a9967d8f2857d

**Comment by rbernon** [comment]
Yes it looks like so, I quickly looked for something like that but couldn't find it. Thanks!

**Comment by smcv** [comment]
This change is making its way through the system and will arrive in a release eventually. Workaround: patch the header in your build environment.

**Comment by smcv** [comment]
This is fixed in Steam Runtime build 0.20200505.0, which comes with the latest stable and beta Steam clients. Please update your Docker container (or schroot, or whatever you are using).

```
eglibc (2.15-0ubuntu10.18+steamrt1.2) scout; urgency=medium

  * d/p/steamrt/BZ-15897-dlfcn-do-not-mark-dlopen-dlclose-as-leaf-fu.diff:
    Backport fix for upstream PR 15897 to fix Wine 5.6 miscompilation.
    (Fixes: ValveSoftware/steam-runtime#250, T20603)
```

**Comment by rbernon** [comment]
Thanks, I can confirm that the dlfcn.h header now has the correct attribute. I didn't retry building Wine 5.6 with Proton yet but I think it should be fine.
